libs/gltfio/src/MaterialGenerator.cpp

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MaterialGenerator.h"
+
+#include <filamat/MaterialBuilder.h>
+
+#include <utils/Log.h>
+
+#include <string>
+
+using namespace filamat;
+using namespace filament;
+using namespace utils;
+
+namespace gltfio {
+namespace details {
+
+bool MaterialGenerator::EqualFn::operator()(const MaterialKey& k1, const MaterialKey& k2) const {
+    return
+        (k1.doubleSided == k2.doubleSided) &&
+        (k1.unlit == k2.unlit) &&
+        (k1.hasVertexColors == k2.hasVertexColors) &&
+        (k1.hasBaseColorTexture == k2.hasBaseColorTexture) &&
+        (k1.hasMetallicRoughnessTexture == k2.hasMetallicRoughnessTexture) &&
+        (k1.hasNormalTexture == k2.hasNormalTexture) &&
+        (k1.hasOcclusionTexture == k2.hasOcclusionTexture) &&
+        (k1.hasEmissiveTexture == k2.hasEmissiveTexture) &&
+        (k1.alphaMode == k2.alphaMode) &&
+        (k1.baseColorUV == k2.baseColorUV) &&
+        (k1.metallicRoughnessUV == k2.metallicRoughnessUV) &&
+        (k1.emissiveUV == k2.emissiveUV) &&
+        (k1.aoUV == k2.aoUV) &&
+        (k1.normalUV == k2.normalUV) &&
+        (k1.alphaMaskThreshold == k2.alphaMaskThreshold);
+}
+
+MaterialGenerator::MaterialGenerator(Engine* engine) : mEngine(engine) {
+    MaterialBuilder::init();
```

Comment:
I recommend calling `MaterialBuilder::shutdown()` in the destructor to free resources

Reply:
Done.